### DISCUSSION TITLE
Offline tile server

### DISCUSSION THREAD

**Discussion by scd31** [dl9rdz/rdz_ttgo_sonde#429] [TTGO firmware]
I'd like to implement an offline tile server in rdz_ttgo_sonde. My thinking is you'd put a directory on the micro SD card, maybe called "tiles" and within it you'd have a folder for each tile set.

While looking into this, I found this discussion: https://github.com/dl9rdz/rdz_ttgo_sonde/discussions/126#discussioncomment-1152484

> Its not so much a problem of code space (well, maybe it is, the SD card file system code might take some space), but more of a RAM issue. The ESP32 web server starts struggeling randomly if you have a single data item of about >10k size to be transferred. I am quite worried that using the ESP32 as something like a tile server will render everything highly unstable.
>
>Also I have not found an easy way for connecting the leaflet map to an offline map, otherwise I would support offline maps in the android app. Offline prediction would be feasible (an an Android phone, not on the TTGO) if you prefetch some weather data.

To me this seems like a solvable problem. The esp32 has an insane amount of CPU/RAM for an embedded chip, and serving static files should be fairly unintensive to do, even for relatively large files.

Assuming I'm able to get this to work in a way that is stable, would it be accepted in a PR?

Also, would a PR to do the flight predictions offline be accepted? My goal is to get this to the point where I can take it plus a phone for my radiosonde hunting - no computer at all.

Thanks for your consideration!

**Comment by dl9rdz** [comment]
Insane amount: Yes and no. With the default Flash size (4M) and layout (2 partitions for OTA + file system) the PlatformIO build used to be at 99% of the space (it is now down to 85.7% (used 1122990 bytes from 1310720 bytes) as I found the reason why pio pulled in the locale support that is not needed).  I'd like to upgrade to newer versions of the runtime environment (which needs more code space) in the future, and having TLS support would also be great for some purposes.

This means:
- Probably, there is not enough space to include such contributions in the default build / binary images compatible with all devices (except the newer S3 ones) on rdzsonde.mooo.com
- I would be happy to accept code with conditional compile (disabled by default with feature macros to enable it in features.h) so interested user can compile on their own (possibly for larger devices with >= 8MB of flash)

What exactly do you mean with offline predictions? The usual predictions use GFS prediction data (world wide data is some gigabytes updated every 6h, ESP32 storage is not insane but ridicously tiny for that). I have been thinking about a web service that actually delivers the right data restricted in time and space for your current location, but that still would be big.  Also, you need elevation data to find out where the ground surface is.  Even if you want to do a simple linear prediction of the flight path ignoring any GFS wind data, you need elevation data. The very coarse grained data (that is also used by tawhiri prediction) is the 15'x15' grid data, which is 2 MB in size. Won't fit...  Ideally you would better use more hi-res data. 

If I had the time I would implement offline prediction for an smartphone app. Ideally combined with some web service that provides efficient download for a regional subset of GFS data (or other, better models) and a high-res elevation model, also regional (at least 3'x'3, some places actually have 1'x1' data available)

**Reply by scd31** [reply]
My thinking was the prediction data would live on the SD card as well. So large amounts of RAM would not be needed for it.

**Reply by dl9rdz** [reply]
Alright, that might fit, but you still need
(a) the code to generate/update that data.
(b) the code to actually do the prediction.
Where would the code for (b) actually execute? I guess not on the ESP32, but on a smartphone? But then why have the data on the ESP32 and not on the phone?

**Reply by scd31** [reply]
(a) would run on a normal computer, and you'd use it to populate the micro SD card

(b) Could probably run fine on the esp32, but if not, you could run it on the phone via front end javascript, and it would just grab the relevant data from the esp32 by making normal HTTP requests. Then you wouldn't need a special app or anything; it would all happen in the browser.
